db/migrate/20170803140326_fix_editionless_change_notes.rb

```diff
@@ -0,0 +1,102 @@
+class FixEditionlessChangeNotes < ActiveRecord::Migration[5.1]
+  disable_ddl_transaction!
+
+  FIRST_PUBLISHED_NOTE = "First published."
+
+  def up
+    removed_change_note_ids = []
+    reassociated_change_note_ids = []
+    document_content_ids = []
+    change_history_from_document_editions = {}
+
+    # ~ 1274 change notes without an association to an edition
+    ChangeNote.where(edition: nil).find_each do |change_note|
+      document = change_note.document
+      change_notes_from_document_editions = ChangeNote.where(edition: document.editions)
+      change_notes_notes_from_document_editions = change_notes_from_document_editions.pluck(:note)
+
+      # Do any of the document editions already have a matching note?
+      if change_notes_notes_from_document_editions.include?(change_note.note)
+        if change_note.destroy!
+          removed_change_note_ids << change_note.id
+          document_content_ids << document.content_id
+          next
+        end
+      end
+
+      # Check if the number of change notes assigned to document editions matches
+      # the number of editions, this change note can't be reassigned so delete it.
+      if change_notes_from_document_editions.count == document.editions.count
+        if change_note.destroy!
+          removed_change_note_ids << change_note.id
+          document_content_ids << document.content_id
+          next
+        end
+      end
+
+      # Check if this is a "First published." note.
+      # If the edition sequence doesn't have one then assign it to the first edition.
+      if change_note.note == FIRST_PUBLISHED_NOTE &&
+        !change_notes_notes_from_document_editions.include?(FIRST_PUBLISHED_NOTE)
+
+        change_note_edition = document.editions.order(:public_updated_at).first
+        unless ChangeNote.find_by(edition: change_note_edition)
+          change_note.edition = change_note_edition
+          if change_note.save!
+            puts "Assigning Edition(id: #{document.editions.first.id}) to ChangeNote(id: #{change_note.id})"
+            reassociated_change_note_ids << change_note.id
+            document_content_ids << document.content_id
+            next
+          end
+        end
+      end
+
+      # Map document.editions change history to a sorted deduplicated array
+      unless change_history_from_document_editions.has_key?(document.id)
+        change_history_from_document_editions[document.id] = mapped_change_history(document)
+      end
+
+      # Cross reference the date and note text of the current change note with a change history
+      # item and match this to an edition (with the same document which doesn't have a change note)
+      # with a public updated date matching both change note and history item.
+      change_note_date = change_note.public_timestamp.to_date
+      matching_editions = document.editions
+        .where("public_updated_at > '#{change_note_date}' and public_updated_at < '#{1.day.since(change_note_date)}'")
```

Comment:
Yes, you're right, there's a chance that the `public_updated_at` time can be the same date at `00:00:00`, sadly it doesn't catch any more records :/